File: duckdb_profiler/duckdb_planbench/io.py
import json, pathlib
from typing import Dict, Any, List, Optional


def save_json(doc: Dict[str, Any], path: str) -> None:
    p = pathlib.Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(doc, indent=2))


def load_json(path: str) -> Dict[str, Any]:
    return json.loads(pathlib.Path(path).read_text())


def draw_tree(
    doc: Dict[str, Any],
    *,
    show_metrics: bool = True,
    show_ms: bool = True,
    show_active_cols: bool = False,
    decimals: int = 3,
) -> str:
    def fmt(n: Dict[str, Any]) -> str:
        op = n.get("op")
        name = n.get("name", "")
        m = n.get("metrics", {})
        cols = (n.get("columns") or {}).get("active") or []
        parts = [op]
        if name:
            parts.append(f"[{name}]")
        if show_metrics:
            t = m.get("elapsed_ms")
            rin = m.get("rows_in")
            rout = m.get("rows_out")
            parts += [
                f"in:{rin if rin is not None else '-'}",
                f"out:{rout if rout is not None else '-'}",
            ]
            if show_ms:
                parts.append(f"ms:{f'{t:.{decimals}f}' if isinstance(t, (int, float)) else '-'}")
            if op == "Join":
                parts.append(f"L_in:{m.get('rows_in_left','-')} R_in:{m.get('rows_in_right','-')}")
        if show_active_cols and cols:
            parts.append(f"cols:{cols}")

        dep_cnt = len(n.get("_deps", [])) if isinstance(n.get("_deps"), list) else 0
        if dep_cnt:
            parts.append(f"deps:{dep_cnt}")

        return " ".join(parts)

    lines: List[str] = []

    def walk(n: Dict[str, Any], pref: str, last: bool):
        lines.append(pref + ("└─ " if last else "├─ ") + fmt(n))
        kids = n.get("children", [])
        for i, ch in enumerate(kids):
            walk(ch, pref + ("   " if last else "│  "), i == len(kids) - 1)

    walk(doc.get("root", {}), "", True)
    qms = doc.get("query_latency_ms")
    if isinstance(qms, (int, float)):
        lines.append(f"\nTotal query latency: {qms:.{decimals}f} ms")
    return "\n".join(lines)


def save_tree_text(
    doc: Dict[str, Any],
    path: str,
    *,
    show_metrics: bool = True,
    show_ms: bool = True,
    show_active_cols: bool = False,
    decimals: int = 3,
) -> None:
    txt = draw_tree(
        doc,
        show_metrics=show_metrics,
        show_ms=show_ms,
        show_active_cols=show_active_cols,
        decimals=decimals,
    )
    p = pathlib.Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(txt)

File: datafusion_profiler/src/io.rs
use anyhow::Result;
use std::path::Path;

use crate::schema::UnifiedDoc;

/// Write the JSON document to path, pretty-printed. Creates directories as needed.
pub fn save_json(doc: &UnifiedDoc, path: &str) -> Result<()> {
    let p = Path::new(path);
    if let Some(par) = p.parent() {
        std::fs::create_dir_all(par)?;
    }
    let txt = serde_json::to_string_pretty(doc)?;
    std::fs::write(p, txt)?;
    Ok(())
}

/// Pretty, single-string ASCII tree based on the Node graph.
pub fn draw_tree(
    doc: &UnifiedDoc,
    show_metrics: bool,
    show_ms: bool,
    show_active_cols: bool,
    decimals: usize,
) -> String {
    fn fmt(
        n: &crate::schema::Node,
        show_metrics: bool,
        show_ms: bool,
        show_active_cols: bool,
        decimals: usize,
    ) -> String {
        let mut parts = vec![n.op.clone(), format!("[{}]", n.name)];
        if show_metrics {
            let m = &n.metrics;
            let rin = m.rows_in.map(|v| v.to_string()).unwrap_or_else(|| "-".into());
            let rout = m.rows_out.map(|v| v.to_string()).unwrap_or_else(|| "-".into());
            let mut metric_line = format!("in:{rin} out:{rout}");
            if show_ms {
                let t = m
                    .elapsed_ms
                    .map(|v| format!("{:.dec$}", v, dec = decimals))
                    .unwrap_or_else(|| "-".into());
                metric_line.push_str(&format!(" ms:{t}"));
            }
            parts.push(metric_line);
            if n.op == "Join" {
                parts.push(format!(
                    "L_in:{} R_in:{}",
                    m.rows_in_left.map(|v| v.to_string()).unwrap_or_else(|| "-".into()),
                    m.rows_in_right.map(|v| v.to_string()).unwrap_or_else(|| "-".into())
                ));
            }
        }
        if show_active_cols && !n.columns.active.is_empty() {
            parts.push(format!("cols:{:?}", n.columns.active));
        }
        parts.join(" ")
    }

    fn walk(
        n: &crate::schema::Node,
        pref: &str,
        last: bool,
        out: &mut Vec<String>,
        show_m: bool,
        show_ms: bool,
        show_c: bool,
        dec: usize,
    ) {
        out.push(format!(
            "{}{} {}",
            pref,
            if last { "└─" } else { "├─" },
            fmt(n, show_m, show_ms, show_c, dec)
        ));
        let kids = &n.children;
        for (i, ch) in kids.iter().enumerate() {
            let next_pref = format!("{}{}", pref, if last { "   " } else { "│  " });
            walk(
                ch,
                &next_pref,
                i == kids.len() - 1,
                out,
                show_m,
                show_ms,
                show_c,
                dec,
            );
        }
    }

    let mut lines = vec![];
    walk(
        &doc.root,
        "",
        true,
        &mut lines,
        show_metrics,
        show_ms,
        show_active_cols,
        decimals,
    );
    if let Some(ms) = doc.query_latency_ms {
        lines.push(format!("\nTotal query latency: {:.dec$} ms", ms, dec = decimals));
    }
    lines.join("\n")
}
